crates/solvers/src/domain/dex/mod.rs

```diff
@@ -120,6 +91,57 @@ impl Swap {
             },
         }
     }
+
+    /// Constructs a single order `solution::Solution` for this swap. Returns
+    /// `None` if the swap is not valid for the specified order.
+    pub fn into_solution(self, order: order::Order) -> Option<solution::Solution> {
+        if !self.matches_order(&order) || !self.respects_price(&order) {
+            return None;
+        }
+
+        let allowance = self.allowance();
+        Some(solution::Solution {
+            prices: solution::ClearingPrices::new([
+                (self.input.token, self.output.amount),
+                (self.output.token, self.input.amount),
+            ]),
+            trades: vec![solution::Trade::Fulfillment(solution::Fulfillment::fill(
+                order,
+            ))],
+            interactions: vec![solution::Interaction::Custom(solution::CustomInteraction {
+                target: self.call.to.0,
+                value: eth::Ether::default(),
+                calldata: self.call.calldata,
+                inputs: vec![self.input],
+                outputs: vec![self.output],
+                internalize: false,
+                allowances: vec![allowance],
+            })],
+        })
+    }
+
+    fn matches_order(&self, order: &order::Order) -> bool {
+        let (swap_amount, order_amount) = match order.side {
+            order::Side::Buy => (self.output.amount, order.buy.amount),
+            order::Side::Sell => (self.input.amount, order.sell.amount),
+        };
+
+        (order.sell.token, order.buy.token) == (self.input.token, self.output.token)
+            && (order.partially_fillable && swap_amount <= order_amount
+                || !order.partially_fillable && swap_amount == order_amount)
+    }
+
+    fn respects_price(&self, order: &order::Order) -> bool {
+        // Note the use of checked multiplication - this is constitant with the
+        // on-chain limit price check.
+        let Some(sell) = order.sell.amount.checked_mul(self.output.amount) else {
+            return false;
```

Comment:
`let else` to return early on `None` is pretty neat. 👍

Reply:
Here is what it looks like with the `match`:
```rust
        match (
            order.sell.amount.checked_mul(self.output.amount),
            order.buy.amount.checked_mul(self.input.amount),
        ) {
            (Some(sell), Some(buy)) => sell >= buy,
            _ => false,
        }
```

I personally found the `let else` way _slightly_ nicer (as it makes the overflow cases look like error branches - which is slightly more accurate in my opinion).

Don't feel strongly though, and happy to change depending on what others find clearer.

Reply:
I had it in mind more like this, but I don't feel strongly about it. Feel free to ignore.
```rust
        let sell = order.sell.amount.checked_mul(self.output.amount);
        let but = order.buy.amount.checked_mul(self.input.amount);
        match (sell, buy) {
            (Some(sell), Some(buy)) => sell >= buy,
            _ => false,
        }
```

Reply:
Can also
```rust
matches!((sell, buy), (Some(sell), Some(buy)) if sell >= buy)
```

Reply:
Awesome!

Reply:
7532fca6b0d943e88845b1d5847bfb52d4230196